pkg/chunkenc/lazy_chunk.go

```diff
@@ -20,21 +21,20 @@ func (c *LazyChunk) getChunk(ctx context.Context) (Chunk, error) {
 	if err != nil {
 		return nil, err
 	}
-
-	c.Chunk = chunks[0]
```

Comment:
this was the so called memory leak. it was retaining the memory until the full expression was evaluated.

Reply:
https://github.com/grafana/loki/pull/713/files#diff-b2f3512576b5e1ebb14517a5c9deccadL31

I was doing it for this "optimisation" :P But I guess you're rewriting this entire bit anyways.